File: src/mysqlcookbook/recipes/transfer/lookup_time.py
#!/usr/bin/python3
# lookup_time.py: Rough test of dictionary versus list lookups.

# - Construct a dictionary with n keys
# - Construct a list with n keys
# - Perform n lookups on each

# This is a very crude test.  Even so, we find the results are often
# striking even for n = 100000.

import time
import sys

n = 100000  # number of keys to use/lookups to perform

mydict = {}
mylist = []

for i in range(1, n+1):
  mydict[i] = 1
  mylist.append(i)

print(time.perf_counter())
print("Performing", n, "dictionary lookups...")
for i in range(1, n+1):
  if i not in mydict:
    sys.exit
print ("Done.")
print(time.perf_counter())
print("Performing", n, "list lookups...")
for i in range(1, n+1):
  if mylist.count(i) <= 0:
    sys.exit
print ("Done.")
print(time.perf_counter())

File: src/mysqlcookbook/recipes/transfer/lookup_time.pl
#!/usr/bin/perl
# lookup_time.pl: Rough test of hash versus array lookups.

# - Construct a hash with n keys
# - Construct an array with n keys
# - Perform n lookups on each

# This is a very crude test.  Even so, I find the results are often
# striking even for n = 1000.

use strict;
use warnings;

my $n = 10000;  # number of keys to use/lookups to perform

my %hash;
my @array;

for my $i (0 .. $n)
{
  $hash{$i} = 1;
  push (@array, $i);
}

print scalar (localtime (time)), "\n";
print "Performing $n hash lookups...\n";
for my $i (0 .. $n)
{
  die "logic error" unless exists ($hash{$i});
}
print "Done.\n";
print scalar (localtime (time)), "\n";
print "Performing $n array lookups...\n";
for my $i (0 .. $n)
{
  die "logic error" unless grep (/^$i$/, @array);
}
print "Done.\n";
print scalar (localtime (time)), "\n";
